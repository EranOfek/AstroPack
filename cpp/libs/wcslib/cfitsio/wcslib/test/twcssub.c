/*============================================================================
  WCSLIB 7.3 - an implementation of the FITS WCS standard.
  Copyright (C) 1995-2020, Mark Calabretta

  This file is part of WCSLIB.

  WCSLIB is free software: you can redistribute it and/or modify it under the
  terms of the GNU Lesser General Public License as published by the Free
  Software Foundation, either version 3 of the License, or (at your option)
  any later version.

  WCSLIB is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for
  more details.

  You should have received a copy of the GNU Lesser General Public License
  along with WCSLIB.  If not, see http://www.gnu.org/licenses.

  Direct correspondence concerning WCSLIB to mark@calabretta.id.au

  Author: Mark Calabretta, Australia Telescope National Facility, CSIRO.
  http://www.atnf.csiro.au/people/Mark.Calabretta
  $Id: twcssub.c,v 7.3.1.2 2020/08/17 11:20:48 mcalabre Exp mcalabre $
*=============================================================================
*
* twcssub tests wcssub() which extracts the coordinate description for a
* subimage from a wcsprm struct.
*
*---------------------------------------------------------------------------*/

#include <stdio.h>
#include <string.h>

#include <wcs.h>
#include <wcserr.h>


// In real life these would be encoded as FITS header keyrecords.
const int NAXIS = 4;
const double CRPIX[4] =  { 1025.0,  64.0, 512.0, 513.0};
const double PC[4][4] = {{    1.1,   0.0,   0.0,   0.0},
                         {    0.0,   1.0,   0.0,   0.0},
                         {    0.0,   0.0,   1.0,   0.1},
                         {    0.0,   0.0,   0.2,   1.0}};
const double CDELT[4] =  {-9.2e-6,  10.0,   1.0,  -1.0};
char CUNIT[4][16] = {"m", "s", "deg", "deg"};
char CTYPE[4][16] = {"WAVE-F2W",  "TIME", "XLAT-SZP",  "XLON-SZP"};

const double CRVAL[4] =  {0.214982042, -2e3, -30.0, 150.0};
const double LONPOLE  = 150.0;
const double LATPOLE  = 999.0;
const double RESTFRQ  =   1.42040575e9;
const double RESTWAV  =   0.0;

char CNAME[4][16] = {"Wavelength", "Time", "Latitude", "Longitude"};

int NPS, NPV;
struct pvcard PV[10];
struct pscard PS[10];


int main()

{
  int axes[4], i, j, nsub, status;
  struct wcsprm wcs, wcsext;
  double *pcij;

  PV[0].i = 1;			// Frequency on axis 1.
  PV[0].m = 1;			// Parameter number 1.
  PV[0].value = -1.0;		// PV1_1.

  PV[1].i = 3;			// Latitude on axis 3.
  PV[1].m = 1;			// Parameter number 1.
  PV[1].value = 2.0;		// PV3_1.

  PV[2].i = 3;			// Latitude on axis 3.
  PV[2].m = 2;			// Parameter number 2.
  PV[2].value = 210.0;		// PV3_2.

  PV[3].i = 3;			// Latitude on axis 3.
  PV[3].m = 3;			// Parameter number 3.
  PV[3].value = 60.0;		// PV3_3.

  NPV = 4;

  PS[0].i = 2;			// Time on axis 2.
  PS[0].m = 1;			// Parameter number 1.
  strcpy(PS[0].value, "UTC");	// PS2_1.

  NPS = 1;

  wcs.flag = -1;
  wcsini(1, NAXIS, &wcs);

  for (j = 0; j < NAXIS; j++) {
    wcs.crpix[j] = CRPIX[j];
  }

  pcij = wcs.pc;
  for (i = 0; i < NAXIS; i++) {
    for (j = 0; j < NAXIS; j++) {
      *(pcij++) = PC[i][j];
    }
  }

  for (i = 0; i < NAXIS; i++) {
    wcs.cdelt[i] = CDELT[i];
  }

  for (i = 0; i < NAXIS; i++) {
    strcpy(wcs.cunit[i], &CUNIT[i][0]);
    strcpy(wcs.ctype[i], &CTYPE[i][0]);
    strcpy(wcs.cname[i], &CNAME[i][0]);
  }

  for (i = 0; i < NAXIS; i++) {
    wcs.crval[i] = CRVAL[i];
  }

  wcs.lonpole = LONPOLE;
  wcs.latpole = LATPOLE;

  wcs.restfrq = RESTFRQ;
  wcs.restwav = RESTWAV;

  wcs.npv = NPV;
  for (i = 0; i < NPV; i++) {
    wcs.pv[i] = PV[i];
  }

  wcs.nps = NPS;
  for (i = 0; i < NPS; i++) {
    wcs.ps[i] = PS[i];
  }


  // Initialize the wcsprm struct.
  wcserr_enable(1);
  (void) wcssetprm(&wcs);

  printf("Testing WCSLIB subimage extraction routine (twcssub.c)\n"
         "------------------------------------------------------\n");
  printf("\nInitial contents of wcsprm struct:\n");
  wcsprt(&wcs);


  // Extract the coordinate description for a subimage and add a new axis.
  nsub = 4;
  wcsext.flag = -1;
  axes[0] = WCSSUB_LONGITUDE;
  axes[1] = WCSSUB_LATITUDE;
  axes[2] = -(WCSSUB_SPECTRAL | WCSSUB_STOKES);
  axes[3] = 0;
  printf("\n\nExtracted contents of wcsprm struct:\n");
  if (wcssub(1, &wcs, &nsub, axes, &wcsext)) {
    wcsperr(&wcsext, "");
  } else if (nsub == 0) {
    printf("None of the requested subimage axes were found.\n");
  } else if (wcssetprm(&wcsext)) {
    wcsperr(&wcsext, "");
  } else {
    wcsprt(&wcsext);
  }


  // Set it up for failure by setting PC1_3 non-zero.
  *(wcs.pc+2) = 1.0;
  nsub = 2;
  axes[0] = 4;
  axes[1] = 3;
  status = wcssub(1, &wcs, &nsub, axes, &wcsext);
  if (status == WCSERR_NON_SEPARABLE) {
    printf("\n\nReceived wcssub status %d as expected for a non-separable "
           "subimage\ncoordinate system.\n", WCSERR_NON_SEPARABLE);
  } else {
    printf("\n\nERROR: expected wcssub status %d for a non-separable "
           "subimage coordinate\nsystem, but received status %d instead.\n",
           WCSERR_NON_SEPARABLE, status);
  }

  wcsfree(&wcs);
  wcsfree(&wcsext);

  return 0;
}
