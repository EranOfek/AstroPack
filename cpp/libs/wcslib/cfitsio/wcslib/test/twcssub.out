Testing WCSLIB subimage extraction routine (twcssub.c)
------------------------------------------------------

Initial contents of wcsprm struct:
       flag: 137
      naxis: 4
      crpix: 0x<address>
               1025.0       64.000       512.00       513.00    
         pc: 0x<address>
    pc[0][]:   1.1000       0.0000       0.0000       0.0000    
    pc[1][]:   0.0000       1.0000       0.0000       0.0000    
    pc[2][]:   0.0000       0.0000       1.0000       0.10000   
    pc[3][]:   0.0000       0.0000       0.20000      1.0000    
      cdelt: 0x<address>
              -9.2000e-06   10.000       1.0000      -1.0000    
      crval: 0x<address>
               0.21498     -2000.0      -30.000       150.00    
      cunit: 0x<address>
             "m"
             "s"
             "deg"
             "deg"
      ctype: 0x<address>
             "WAVE-F2W"
             "TIME"
             "XLAT-SZP"
             "XLON-SZP"
    lonpole: 150.000000
    latpole: -30.000000
    restfrq: 1420405750.000000
    restwav: 0.000000
        npv: 4
     npvmax: 64
         pv: 0x<address>
               1   1  -1.0000    
               3   1   2.0000    
               3   2   210.00    
               3   3   60.000    
        nps: 1
     npsmax: 8
         ps: 0x<address>
               2   1  UTC
         cd: 0x<address>
    cd[0][]:   0.0000       0.0000       0.0000       0.0000    
    cd[1][]:   0.0000       0.0000       0.0000       0.0000    
    cd[2][]:   0.0000       0.0000       0.0000       0.0000    
    cd[3][]:   0.0000       0.0000       0.0000       0.0000    
      crota: 0x<address>
               0.0000       0.0000       0.0000       0.0000    
     altlin: 0
     velref: 0
        alt: ' '
     colnum: 0
      colax: 0x<address>
                 0      0      0      0
      cname: 0x<address>
             "Wavelength"
             "Time"
             "Latitude"
             "Longitude"
      crder: 0x<address>
               UNDEFINED    UNDEFINED    UNDEFINED    UNDEFINED
      csyer: 0x<address>
               UNDEFINED    UNDEFINED    UNDEFINED    UNDEFINED
      czphs: 0x<address>
               UNDEFINED    UNDEFINED    UNDEFINED    UNDEFINED
      cperi: 0x<address>
               UNDEFINED    UNDEFINED    UNDEFINED    UNDEFINED
    wcsname: UNDEFINED
    timesys: UNDEFINED
    trefpos: UNDEFINED
    trefdir: UNDEFINED
    plephem: UNDEFINED
   timeunit: UNDEFINED
    dateref: UNDEFINED
     mjdref:      0.000000000     0.000000000
   timeoffs: UNDEFINED
    dateobs: UNDEFINED
    datebeg: UNDEFINED
    dateavg: UNDEFINED
    dateend: UNDEFINED
     mjdobs: UNDEFINED
     mjdbeg: UNDEFINED
     mjdavg: UNDEFINED
     mjdend: UNDEFINED
     jepoch: UNDEFINED
     bepoch: UNDEFINED
     tstart: UNDEFINED
      tstop: UNDEFINED
    xposure: UNDEFINED
    telapse: UNDEFINED
    timsyer: UNDEFINED
    timrder: UNDEFINED
    timedel: UNDEFINED
   timepixr: UNDEFINED
     obsgeo:        UNDEFINED       UNDEFINED       UNDEFINED
                    UNDEFINED       UNDEFINED       UNDEFINED
   obsorbit: UNDEFINED
    radesys: UNDEFINED
    equinox: UNDEFINED
    specsys: UNDEFINED
    ssysobs: UNDEFINED
    velosys: UNDEFINED
    zsource: UNDEFINED
    ssyssrc: UNDEFINED
    velangl: UNDEFINED
        aux: 0x0
       ntab: 0
        tab: 0x0
       nwtb: 0
        wtb: 0x0
      types: 0x<address>
            3300    0 2201 2200
     lngtyp: "XLON"
     lattyp: "XLAT"
        lng: 3
        lat: 2
       spec: 0
   cubeface: -1
        err: 0x0
        lin: (see below)
        cel: (see below)
        spc: (see below)
     m_flag: 137
    m_naxis: 4
    m_crpix: 0x<address>  (= crpix)
       m_pc: 0x<address>  (= pc)
    m_cdelt: 0x<address>  (= cdelt)
    m_crval: 0x<address>  (= crval)
    m_cunit: 0x<address>  (= cunit)
    m_ctype: 0x<address>  (= ctype)
       m_pv: 0x<address>  (= pv)
       m_ps: 0x<address>  (= ps)
       m_cd: 0x<address>  (= cd)
    m_crota: 0x<address>  (= crota)

    m_colax: 0x<address>  (= colax)
    m_cname: 0x<address>  (= cname)
    m_crder: 0x<address>  (= crder)
    m_csyer: 0x<address>  (= csyer)
    m_czphs: 0x<address>  (= czphs)
    m_cperi: 0x<address>  (= cperi)
      m_aux: 0x0  (= aux)
      m_tab: 0x0  (= tab)
      m_wtb: 0x0  (= wtb)

   lin.*
       flag: 137
      naxis: 4
      crpix: 0x<address>
               1025.0       64.000       512.00       513.00    
         pc: 0x<address>
    pc[0][]:   1.1000       0.0000       0.0000       0.0000    
    pc[1][]:   0.0000       1.0000       0.0000       0.0000    
    pc[2][]:   0.0000       0.0000       1.0000       0.10000   
    pc[3][]:   0.0000       0.0000       0.20000      1.0000    
      cdelt: 0x<address>
              -9.2000e-06   10.000       1.0000      -1.0000    
     dispre: 0x0
     disseq: 0x0
piximg[0][]:  -1.0120e-05  -0.0000      -0.0000      -0.0000    
piximg[1][]:   0.0000       10.000       0.0000       0.0000    
piximg[2][]:   0.0000       0.0000       1.0000       0.10000   
piximg[3][]:  -0.0000      -0.0000      -0.20000     -1.0000    
imgpix[0][]:  -98814.      -0.0000      -0.0000      -0.0000    
imgpix[1][]:   0.0000       0.10000      0.0000       0.0000    
imgpix[2][]:   0.0000       0.0000       1.0204       0.10204   
imgpix[3][]:  -0.0000      -0.0000      -0.20408     -1.0204    
    i_naxis: 4
      unity: 0
     affine: 1
     simple: 0
        err: 0x0
     tmpcrd: 0x<address>
     m_flag: 0
    m_naxis: 0
    m_crpix: 0x0
       m_pc: 0x0
    m_cdelt: 0x0
   m_dispre: 0x0
   m_disseq: 0x0

   cel.*
      flag: 137
     offset: 0
       phi0:  0.000000
     theta0: 90.000000
        ref:   150.00      -30.000       150.00      -30.000    
        prj: (see below)
      euler:   150.00       120.00       150.00      -0.50000      0.86603   
    latpreq: 0 (not required)
     isolat: 0
        err: 0x0

   prj.*
       flag: 102
       code: "SZP"
         r0: 57.295780
         pv: (0)
              2.0000       210.00       60.000    
       phi0:  0.000000
     theta0: 90.000000
     bounds: 7

       name: "slant zenithal perspective"
   category: 1 (zenithal)
    pvrange: 103
  simplezen: 0
  equiareal: 0
  conformal: 0
     global: 0
  divergent: 0
         x0: 0.000000
         y0: 0.000000
        err: 0x0
        w[]:   0.017453     0.50000     -0.86603      2.7321       28.648    
              -49.620       156.53       3.7321      -90.000       0.0000    
          m: 0
          n: 0
     prjx2s: 0x<address>
     prjs2x: 0x<address>

   spc.*
       flag: 110
       type: "WAVE"
       code: "F2W"
      crval:  0.21498   
    restfrq: 1420405750.000000
    restwav: 0.000000
         pv: (not used)
          w:   0.0000       1.3945e+09  -6.4866e+09  (remainder unused)
    isGrism: 0
        err: 0x0
     spxX2P: 0x<address>
     spxP2S: 0x0
     spxS2P: 0x0
     spxP2X: 0x<address>


Extracted contents of wcsprm struct:
       flag: 137
      naxis: 4
      crpix: 0x<address>
               513.00       512.00       64.000       0.0000    
         pc: 0x<address>
    pc[0][]:   1.0000       0.20000      0.0000       0.0000    
    pc[1][]:   0.10000      1.0000       0.0000       0.0000    
    pc[2][]:   0.0000       0.0000       1.0000       0.0000    
    pc[3][]:   0.0000       0.0000       0.0000       1.0000    
      cdelt: 0x<address>
              -1.0000       1.0000       10.000       1.0000    
      crval: 0x<address>
               150.00      -30.000      -2000.0       0.0000    
      cunit: 0x<address>
             "deg"
             "deg"
             "s"
             ""
      ctype: 0x<address>
             "XLON-SZP"
             "XLAT-SZP"
             "TIME"
             ""
    lonpole: 150.000000
    latpole: -30.000000
    restfrq: 1420405750.000000
    restwav: 0.000000
        npv: 3
     npvmax: 3
         pv: 0x<address>
               2   1   2.0000    
               2   2   210.00    
               2   3   60.000    
        nps: 1
     npsmax: 1
         ps: 0x<address>
               3   1  UTC
         cd: 0x<address>
    cd[0][]:   0.0000       0.0000       0.0000       0.0000    
    cd[1][]:   0.0000       0.0000       0.0000       0.0000    
    cd[2][]:   0.0000       0.0000       0.0000       0.0000    
    cd[3][]:   0.0000       0.0000       0.0000       0.0000    
      crota: 0x<address>
               0.0000       0.0000       0.0000       0.0000    
     altlin: 0
     velref: 0
        alt: ' '
     colnum: 0
      colax: 0x<address>
                 0      0      0      0
      cname: 0x<address>
             "Longitude"
             "Latitude"
             "Time"
             UNDEFINED
      crder: 0x<address>
               UNDEFINED    UNDEFINED    UNDEFINED    UNDEFINED
      csyer: 0x<address>
               UNDEFINED    UNDEFINED    UNDEFINED    UNDEFINED
      czphs: 0x<address>
               UNDEFINED    UNDEFINED    UNDEFINED    UNDEFINED
      cperi: 0x<address>
               UNDEFINED    UNDEFINED    UNDEFINED    UNDEFINED
    wcsname: UNDEFINED
    timesys: UNDEFINED
    trefpos: UNDEFINED
    trefdir: UNDEFINED
    plephem: UNDEFINED
   timeunit: UNDEFINED
    dateref: UNDEFINED
     mjdref:      0.000000000     0.000000000
   timeoffs: UNDEFINED
    dateobs: UNDEFINED
    datebeg: UNDEFINED
    dateavg: UNDEFINED
    dateend: UNDEFINED
     mjdobs: UNDEFINED
     mjdbeg: UNDEFINED
     mjdavg: UNDEFINED
     mjdend: UNDEFINED
     jepoch: UNDEFINED
     bepoch: UNDEFINED
     tstart: UNDEFINED
      tstop: UNDEFINED
    xposure: UNDEFINED
    telapse: UNDEFINED
    timsyer: UNDEFINED
    timrder: UNDEFINED
    timedel: UNDEFINED
   timepixr: UNDEFINED
     obsgeo:        UNDEFINED       UNDEFINED       UNDEFINED
                    UNDEFINED       UNDEFINED       UNDEFINED
   obsorbit: UNDEFINED
    radesys: UNDEFINED
    equinox: UNDEFINED
    specsys: UNDEFINED
    ssysobs: UNDEFINED
    velosys: UNDEFINED
    zsource: UNDEFINED
    ssyssrc: UNDEFINED
    velangl: UNDEFINED
        aux: 0x0
       ntab: 0
        tab: 0x0
       nwtb: 0
        wtb: 0x0
      types: 0x<address>
            2200 2201    0    0
     lngtyp: "XLON"
     lattyp: "XLAT"
        lng: 0
        lat: 1
       spec: -1
   cubeface: -1
        err: 0x0
        lin: (see below)
        cel: (see below)
        spc: (see below)
     m_flag: 137
    m_naxis: 4
    m_crpix: 0x<address>  (= crpix)
       m_pc: 0x<address>  (= pc)
    m_cdelt: 0x<address>  (= cdelt)
    m_crval: 0x<address>  (= crval)
    m_cunit: 0x<address>  (= cunit)
    m_ctype: 0x<address>  (= ctype)
       m_pv: 0x<address>  (= pv)
       m_ps: 0x<address>  (= ps)
       m_cd: 0x<address>  (= cd)
    m_crota: 0x<address>  (= crota)

    m_colax: 0x<address>  (= colax)
    m_cname: 0x<address>  (= cname)
    m_crder: 0x<address>  (= crder)
    m_csyer: 0x<address>  (= csyer)
    m_czphs: 0x<address>  (= czphs)
    m_cperi: 0x<address>  (= cperi)
      m_aux: 0x0  (= aux)
      m_tab: 0x0  (= tab)
      m_wtb: 0x0  (= wtb)

   lin.*
       flag: 137
      naxis: 4
      crpix: 0x<address>
               513.00       512.00       64.000       0.0000    
         pc: 0x<address>
    pc[0][]:   1.0000       0.20000      0.0000       0.0000    
    pc[1][]:   0.10000      1.0000       0.0000       0.0000    
    pc[2][]:   0.0000       0.0000       1.0000       0.0000    
    pc[3][]:   0.0000       0.0000       0.0000       1.0000    
      cdelt: 0x<address>
              -1.0000       1.0000       10.000       1.0000    
     dispre: 0x0
     disseq: 0x0
piximg[0][]:  -1.0000      -0.20000     -0.0000      -0.0000    
piximg[1][]:   0.10000      1.0000       0.0000       0.0000    
piximg[2][]:   0.0000       0.0000       10.000       0.0000    
piximg[3][]:   0.0000       0.0000       0.0000       1.0000    
imgpix[0][]:  -1.0204      -0.20408     -0.0000      -0.0000    
imgpix[1][]:   0.10204      1.0204       0.0000       0.0000    
imgpix[2][]:   0.0000       0.0000       0.10000      0.0000    
imgpix[3][]:   0.0000       0.0000       0.0000       1.0000    
    i_naxis: 4
      unity: 0
     affine: 1
     simple: 0
        err: 0x0
     tmpcrd: 0x<address>
     m_flag: 0
    m_naxis: 0
    m_crpix: 0x0
       m_pc: 0x0
    m_cdelt: 0x0
   m_dispre: 0x0
   m_disseq: 0x0

   cel.*
      flag: 137
     offset: 0
       phi0:  0.000000
     theta0: 90.000000
        ref:   150.00      -30.000       150.00      -30.000    
        prj: (see below)
      euler:   150.00       120.00       150.00      -0.50000      0.86603   
    latpreq: 0 (not required)
     isolat: 0
        err: 0x0

   prj.*
       flag: 102
       code: "SZP"
         r0: 57.295780
         pv: (0)
              2.0000       210.00       60.000    
       phi0:  0.000000
     theta0: 90.000000
     bounds: 7

       name: "slant zenithal perspective"
   category: 1 (zenithal)
    pvrange: 103
  simplezen: 0
  equiareal: 0
  conformal: 0
     global: 0
  divergent: 0
         x0: 0.000000
         y0: 0.000000
        err: 0x0
        w[]:   0.017453     0.50000     -0.86603      2.7321       28.648    
              -49.620       156.53       3.7321      -90.000       0.0000    
          m: 0
          n: 0
     prjx2s: 0x<address>
     prjs2x: 0x<address>

   spc.*
       flag: 0
       type: "    "
       code: "   "
      crval: UNDEFINED
    restfrq: 0.000000
    restwav: 0.000000
         pv: (not used)
          w:   0.0000       0.0000       0.0000      (remainder unused)
    isGrism: 0
        err: 0x0
     spxX2P: 0x0
     spxP2S: 0x0
     spxS2P: 0x0
     spxP2X: 0x0


Received wcssub status 13 as expected for a non-separable subimage
coordinate system.
